FractionalMaxPool3d with no_batch_dim support

george-qi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69732

Differential Revision: [D33280090](https://our.internmc.facebook.com/intern/diff/D33280090)

george-qi committed:
FractionalMaxPool with no_batch_dim support

[ghstack-poisoned]

jbschlosser commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -849,14 +849,15 @@ inline std::tuple<Tensor, Tensor> fractional_max_pool3d_with_indices(
   Tensor _random_samples_ = _random_samples;
   if (!_random_samples_.defined()) {
-    _random_samples_ = torch::rand({input.size(0), input.size(1), 3}, torch::TensorOptions().dtype(input.dtype()).device(input.device()));
+    auto n_batch = 1 ? input.dim() == 4 : input.size(0);
```
I think you want `auto n_batch = input.dim() == 4 ? 1 : input.size(0)`, no?

george-qi committed:
Update on "FractionalMaxPool with no_batch_dim support"




[ghstack-poisoned]

george-qi committed:
Update on "FractionalMaxPool with no_batch_dim support"




[ghstack-poisoned]

george-qi committed:
Update on "FractionalMaxPool3d with no_batch_dim support"




[ghstack-poisoned]

george-qi committed:
Update on "FractionalMaxPool3d with no_batch_dim support"




[ghstack-poisoned]

george-qi committed:
Update on "FractionalMaxPool3d with no_batch_dim support"




[ghstack-poisoned]

george-qi committed:
Update on "FractionalMaxPool3d with no_batch_dim support"




[ghstack-poisoned]

zou3519 commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -849,14 +849,15 @@ inline std::tuple<Tensor, Tensor> fractional_max_pool3d_with_indices(
-    output_size_ = {(int64_t)(input.sizes()[2] * (*output_ratio.value())[0]),
-                    (int64_t)(input.sizes()[3] * (*output_ratio.value())[1]),
-                    (int64_t)(input.sizes()[4] * (*output_ratio.value())[2])};
+    output_size_ = {(int64_t)(input.sizes()[-3] * (*output_ratio.value())[0]),
+                    (int64_t)(input.sizes()[-2] * (*output_ratio.value())[1]),
+                    (int64_t)(input.sizes()[-1] * (*output_ratio.value())[2])};
```
Indexing into IntArrayRef doesn't wrap around. input.sizes()[-1] actually tries to index `input.sizes()[MAX_INT - 1]` which is definitely not what you wanted

zou3519 replied:
input.size(-1) should do the trick though

george-qi committed:
Update on "FractionalMaxPool3d with no_batch_dim support"




[ghstack-poisoned]

george-qi commented:
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33280090).

jbschlosser commented on torch/csrc/api/include/torch/nn/functional/pooling.h:
```diff
@@ -4,6 +4,7 @@
 #include <torch/nn/functional/activation.h>
 #include <torch/nn/options/pooling.h>
 #include <torch/nn/modules/utils.h>
+#include <iostream>
```
nit: was this added for debugging? if so, it should be removed

jbschlosser reviewed (commented):
Looking good! Just one tiny nit

george-qi committed:
Update on "FractionalMaxPool3d with no_batch_dim support"


Differential Revision: [D33280090](https://our.internmc.facebook.com/intern/diff/D33280090)

[ghstack-poisoned]

george-qi commented:
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33280090).

jbschlosser reviewed (approved):
LGTM!
